src/syntax/FStar.Syntax.Syntax.fsti

```diff
@@ -429,39 +449,41 @@ type wp_eff_combinators = {
 (*
  * Layered effects combinators
  *
- * All of these are pairs of type schemes,
+ * All of these have pairs of type schemes,
  *   where the first component is the term ts and the second component is the type ts
  *
  * Before typechecking the effect declaration, the second component is a dummy ts
  *   In other words, desugaring sets the first component only, and typechecker then fills up the second one
  *
- * Similarly the base effect name is also "" after desugaring, and is set by the typechecker
+ * Additionally, bind, subcomp, and if_then_else have a combinator kind,
+ *   this is also set to None in desugaring and set during typechecking the effect
  *)
 type layered_eff_combinators = {
   l_repr         : (tscheme * tscheme);
   l_return       : (tscheme * tscheme);
-  l_bind         : (tscheme * tscheme);
-  l_subcomp      : (tscheme * tscheme);
-  l_if_then_else : (tscheme * tscheme);
-
+  l_bind         : (tscheme * tscheme * option indexed_effect_combinator_kind);
+  l_subcomp      : (tscheme * tscheme * option indexed_effect_combinator_kind);
+  l_if_then_else : (tscheme * tscheme * option indexed_effect_combinator_kind);
 }
 
-
 type eff_combinators =
   | Primitive_eff of wp_eff_combinators
   | DM4F_eff of wp_eff_combinators
   | Layered_eff of layered_eff_combinators
 
+type effect_signature =
+  | Layered_eff_sig of int & tscheme  // (n, ts) where n is the number of effect parameters (all upfront) in the effect signature
```

Comment:
Indexed effect signature also maintains the number of binders in the signature that are effect parameters. This is determined by an attribute `effect_param` marked in the effect signature definition. `ToSyntax` fills in this int in the AST, and strips off the `effect_param` attribute, so it is not used in the rest of the typechecker afterwards.